retrieve-logs/fetch_missing_matomo_requests.py

```diff
@@ -0,0 +1,110 @@
+from datetime import datetime, timedelta, timezone
+import os
+import boto3
+import time
+from _decimal import Decimal
+
+NUM_OF_DAYS = 'NUM_OF_DAYS'
+START_DATE = 'START_DATE'
+
+
+def print_unset_env_variable_error_and_exit(environment_variable):
+    print(environment_variable, " environment variable is not set.")
+    exit(1)
+
+
+def validate_environment_variables():
+    if os.getenv(START_DATE) is None:
+        print_unset_env_variable_error_and_exit(START_DATE)
+    if os.getenv(NUM_OF_DAYS) is None:
+        print_unset_env_variable_error_and_exit(NUM_OF_DAYS)
+
+
+def get_start_date():
+    try:
+        return datetime.strptime(os.getenv(START_DATE), '%Y-%m-%dT%H:%M:%S%z')
+    except ValueError:
+        print("START_DATE has an invalid date and time format. It should be in %Y-%m-%dT%H:%M:%S%z")
+        exit(1)
+
+
+def get_number_of_days():
+    try:
+        return int(os.getenv(NUM_OF_DAYS))
+    except ValueError:
+        print("NUM_OF_DAYS has an invalid format. It should be in integers only")
+        exit(1)
+
+
+def wait_for_the_query_to_complete(response):
+    queryId = response['queryId']
+    status = 'Running'
+    while status != 'Complete':
+        response = client.get_query_results(queryId=queryId)
+        status = response['status']
+        time.sleep(1)
+    return response
+
+
+def run_query(start_timestamp, end_timestamp):
+    return client.start_query(
+        logGroupName='matomo',
+        startTime=int(start_timestamp.timestamp() * 1000),
+        endTime=int(end_timestamp.timestamp() * 1000),
+        queryString=
+        """fields @message
+        | sort @timestamp asc
+        | filter @logStream like /matomo-nginx/
+        | filter status!='200'
+        | filter status!='204'
+        | filter user_agent!='ELB-HealthChecker/2.0'
+        | filter path like /idsite=1/
+        | filter path like /rec=1/""",
+        limit=10000
+    )
+
+
+def write_requests_to_a_file(response, start_date, end_date, output_filename):
+    with open(output_filename, 'a+') as f:
+        for message in response['results']:
+            for message in message:
```

Comment:
Surprised the linter didn't complain about this variable hiding!?

Reply:
🤔 Good point. I won't address this here, but it would definitely be worth changing this.

Reply:
Changed in https://github.com/alphagov/verify-matomo-utils/pull/3/commits/867f854d8cd6d02fc51663d9cbe99a4dd15475d3